dbms/src/Storages/DeltaMerge/DeltaMergeStore.h

```diff
@@ -483,11 +458,81 @@ class DeltaMergeStore : private boost::noncopyable
      */
     void checkSegmentUpdate(const DMContextPtr & context, const SegmentPtr & segment, ThreadType thread_type);
 
+#pragma region Internal Segment Operations
+
+#pragma region Segment Split
+
+    enum class SegmentSplitReason
+    {
+        ForegroundWrite,
+        Background,
+        IngestBySplit,
+    };
+
+    /**
+     * Note: This enum simply shadows Segment::SplitMode without introducing the whole Segment into this header.
+     */
+    enum class SegmentSplitMode
+    {
+        /**
+         * Split according to settings.
+         *
+         * If logical split is allowed in the settings, logical split will be tried first.
+         * Logical split may fall back to physical split when calculating split point failed.
+         */
+        Auto,
+
+        /**
+         * Do logical split. If split point is not specified and cannot be calculated out,
+         * the split will fail.
+         */
+        Logical,
+
+        /**
+         * Do physical split.
+         */
+        Physical,
+    };
+
+    static std::string toString(SegmentSplitMode mode)
+    {
+        switch (mode)
+        {
+        case SegmentSplitMode::Auto:
+            return "Auto";
+        case SegmentSplitMode::Logical:
+            return "Logical";
+        case SegmentSplitMode::Physical:
+            return "Physical";
+        default:
+            return "Unknown";
+        }
+    }
+
+    static std::string toString(SegmentSplitReason reason)
+    {
+        switch (reason)
+        {
+        case SegmentSplitReason::ForegroundWrite:
+            return "ForegroundWrite";
+        case SegmentSplitReason::Background:
+            return "Background";
+        case SegmentSplitReason::IngestBySplit:
+            return "IngestBySplit";
+        default:
+            return "Unknown";
+        }
+    }
+
     /**
      * Split the segment into two.
      * After splitting, the segment will be abandoned (with `segment->hasAbandoned() == true`) and the new two segments will be returned.
+     *
+     * When `opt_split_at` is not specified, this function will try to find a mid point for splitting, and may lead to failures.
      */
-    SegmentPair segmentSplit(DMContext & dm_context, const SegmentPtr & segment, bool is_foreground);
+    SegmentPair segmentSplit(DMContext & dm_context, const SegmentPtr & segment, SegmentSplitReason reason, std::optional<RowKeyValue> opt_split_at = std::nullopt, SegmentSplitMode opt_split_mode = SegmentSplitMode::Auto);
+
+#pragma endregion Segment Split
```

Comment:
I'm sorry what is it used for?

Reply:
It triggers a collapseable block in IDEs like VS, Intellij Idea and VSCode via extensions, like this, so that we could better organize large files:

<img width="559" alt="image" src="https://user-images.githubusercontent.com/1916485/190312948-d91b0297-8411-4a45-a7dc-654e2033aca1.png">

<img width="503" alt="image" src="https://user-images.githubusercontent.com/1916485/190312980-7c3789ea-eda7-413b-8e02-08c36f7e745a.png">


However I just figured out that GCC will generate a warning for this instruction (although I don't have an environment to verify it):
https://gcc.gnu.org/bugzilla//show_bug.cgi?id=85487

So maybe I should drop it when we are still considering GCC compatibility (in the README)?